src/mlpack/methods/softmax_regression/softmax_regression.hpp

```diff
@@ -98,15 +96,25 @@ class SoftmaxRegression
                     const bool fitIntercept = false);
 
   /**
-   * Construct the softmax regression model with the given training data. This
-   * will train the model. This overload takes an already instantiated optimizer
-   * and uses it to train the model. The optimizer should hold an instantiated
-   * SoftmaxRegressionFunction object for the function to operate upon. This
-   * option should be preferred when the optimizer options are to be changed.
+   * Construct the SoftmaxRegression class with the provided data and labels.
+   * This will train the model. Optionally, the parameter 'lambda' can be
+   * passed, which controls the amount of L2-regularization in the objective
+   * function. By default, the model takes a small value.
    *
-   * @param optimizer Instantiated optimizer with instantiated error function.
+   * @param data Input training features. Each column associate with one sample
+   * @param labels Labels associated with the feature data.
+   * @param inputSize Size of the input feature vector.
+   * @param numClasses Number of classes for classification.
+   * @param optimizer Desired optimizer.
+   * @param lambda L2-regularization constant.
+   * @param fitIntercept add intercept term or not.
    */
-  SoftmaxRegression(OptimizerType<SoftmaxRegressionFunction>& optimizer);
+  SoftmaxRegression(const arma::mat& data,
+                    const arma::Row<size_t>& labels,
+                    const size_t numClasses,
+                    OptimizerType& optimizer,
```

Comment:
Related to #1021, I think we could just use `OptimizerType optimizer = OptimizerType()` here and use that as the last argument to the other constructor.  But I don't see the need to refactor everything here, we can save that for later.